ESA BIOMASS Simulated data access notebook

omshinde opened the pull request:
https://github.com/NASA-IMPACT/active-maap-sprint/issues/1308

Illustrates ESA BIOMASS simulated data access using ESA MAAP long-lasting (offline token) and short-lived (data access token) tokens.

omshinde committed:
ESA BIOMASS Simulated data access notebook

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>rephrase run this notebook to To access and run this tutorial within MAAP's Algorithm Development Environment (ADE), please refer to the&nbsp;<a href="https://docs.maap-project.org/en/latest/getting_started/getting_started.html" target="_blank">"Getting started with the MAAP"</a>&nbsp;section of our documentation.</p><p>Disclaimer: it is highly recommended to run a tutorial within MAAP's ADE, which already includes packages specific to MAAP, such as maap-py. Running the tutorial outside of the MAAP ADE may lead to errors.</p><p> </p><p>Ferrante, Francesco (SERCO) (comma)</p><p> </p><p>maybe tag ESA_CCI_V5_Token_Access page link as additional resource. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571134'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

omshinde replied:
<p>Done. </p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>in the ESA_CCI_V5_Token_Access the Token paste is hidden so it won’t show in outputs/screenshots.</p><p>Token is saved in a user config folder </p><p>Saved token file is locked down  so only you can read it </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571136'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='5'/>

omshinde replied:
<p>Deleted the print statement.</p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #10.    # Quick sanity check: list a few collections (should include BiomassSimulated variants)
</pre><p>maybe call this part "Testing Data Access"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571138'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>rename to "Discover tiles via ESA STAC"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571139'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='9'/>

omshinde replied:
<p>I think it's better to keep as "Testing data Access" following other notebooks. </p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #4.    print("Total items:", items)
</pre><p>maybe just print len(items) to make output less noisy.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571140'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

omshinde replied:
<p>Done</p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>open via rasterio with auth headers passed through GDAL, then read insted of full download even when the file is a COG.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571141'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

omshinde replied:
<p>Done. </p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #14.    import os, numpy as np, requests, matplotlib.pyplot as plt, rasterio
</pre><p>move to top under imports.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571144'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

omshinde replied:
<p>Done</p>

HarshiniGirish commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Simulated_Data_Access.ipynb:
```diff
@@ -0,0 +1,524 @@
+{
```
<p>successfully. typo</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2625571148'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='17'/>

omshinde replied:
<p>updated. </p>

HarshiniGirish commented:
I made a few minor suggestions, but overall everything looks good. Thank you! @omshinde

omshinde committed:
Adding ESA BIOMASS Level1A and simulated data access notebook

omshinde commented:
Thanks @HarshiniGirish .. I requested re-review after addressing all your comments. Also, pushed another notebook for ESA BIOMASS Level1A data product.

wildintellect commented:
Should we change to focus on the L1b or L2a which are more likely to be COGs?

List from Cristiano
>  BiomassLevel1a - an SLC (ignore the Sx_SCS__1M data).
> BiomassLevel1b - and GRD like, ARD product
> BiomassLevel1c - co-registered SLC, only after complete swaths cycles, Swath-1 should be coming soon.
> BiomassLevel2a - Forest retrieval products,  generated from L1C.

hrodmn commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Data_Access.ipynb:
```diff
@@ -0,0 +1,556 @@
+{
```
<p>The ESA STAC API is not protected by any auth scheme so we don't need to provide the token header here!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2664867334'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

omshinde replied:
<p>Good catch. Thanks. </p>

hrodmn replied:
<p>Since we aren't using the auth headers in the STAC API requests I would move the <code>headers = {...}</code> definition to the first cell where we do need it.</p>

hrodmn commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Data_Access.ipynb:
```diff
@@ -0,0 +1,556 @@
+{
```
<p>The ESA STAC API has the <code>filter</code> extension enabled which means we can apply the product type filter here instead of sorting through the item ids later. Unfortunately they only enabled it for the GET request model so we need to set method="GET". The <a href="https://catalog.maap.eo.esa.int/doc/stac.html#search-with-filter" target="_blank">ESA STAC API docs</a>  provide a similar example.</p><p> </p>

<pre class="ql-syntax" spellcheck="false">search = stac.search(
&nbsp; &nbsp; collections=["BiomassLevel1a"],
&nbsp; &nbsp; datetime=time_day,
&nbsp; &nbsp; limit=1000,
&nbsp; &nbsp; filter="productType='S2_SCS__1S'",
&nbsp; &nbsp; method="GET",
)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2664867340'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

omshinde replied:
<p>Thanks! Updated it to filter it using the <code>filter</code> extension.  </p>

hrodmn commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Data_Access.ipynb:
```diff
@@ -0,0 +1,556 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #30.    r = requests.get(asset_href, headers=headers, stream=True, timeout=180)
</pre><p>code that downloads the actual assets should be the only place where we need the authorization header</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2664867342'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

omshinde replied:
<p>thanks</p>

hrodmn commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Data_Access.ipynb:
```diff
@@ -0,0 +1,556 @@
+{
```
<p>It would be helpful to provide suggestions for how to get authenticated here. I tried following the link and logging in with my EDL credentials but got bounced with this message "The ID token issued by the identity provider does not match the configured essential claim. Please contact your administrator."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2665548743'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

omshinde replied:
<p>This seems to be an issue under discussion with ESA team for NASA EDL authentication. </p>

omshinde committed:
Update notebook to access L1B product

omshinde commented:
> Should we change to focus on the L1b or L2a which are more likely to be COGs?

Thanks for your comment @wildintellect. I was using L1a which is also COG but I updated it to L1b which seems to be ARD.

omshinde commented:
@hrodmn Addressed all of your comments [here](https://github.com/MAAP-Project/maap-documentation/pull/543/commits/53d1160250386dfd8629e8463b19ebb69d261b38) Requesting re-review

omshinde committed:
Fix minor typo

omshinde committed:
Add indexing details

omshinde committed:
Fix credentials.txt path

omshinde committed:
minor fix

hrodmn commented on docs/source/science/ESA_BIOMASS/ESA_BIOMASS_Data_Access.ipynb:
```diff
@@ -0,0 +1,551 @@
+{
```
<p>I think we should steer users to using this approach (accessing the asset remotely over HTTP instead of downloading to disk) by default. Consider combining this example with the visualization example so there is an example that shows how to do the visualization task in a cloud-native way.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2681935783'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

hrodmn replied:
<p>On the Hub we can use the collection free-text feature. This works on the ADE too but you need to install pystac-client&gt;=0.8.4</p>

<pre class="ql-syntax" spellcheck="false">search = stac.collection_search(q="BiomassLevel1b")
_cols = search.collection_list()


print(f"Found {len(_cols)} collections containing 'BiomassLevel1b':")
for c in _cols:
&nbsp; &nbsp; print(f"&nbsp; - {c.id} :: {c.title}")
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2681965528'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='8'/>

hrodmn replied:
<pre class="ql-syntax" spellcheck="false">Line #5.    items = search.items()
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2682027330'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

hrodmn replied:
<pre class="ql-syntax" spellcheck="false">Line #6.    print("Total items:", search.matched())
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2682027332'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

hrodmn replied:
<pre class="ql-syntax" spellcheck="false">Line #2.    s3_item = next(items)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MAAP-Project/maap-documentation/pull/543/discussion/#comment-2682027337'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

hrodmn commented:
@omshinde I made a few suggestions on a cleaner way to search for the biomass collection and how to use pystac-client more efficiently during the item search step. The structural comment about combining the remote access and granule visualization examples is not a blocker but something that could be worth changing if you have time this week.

HarshiniGirish committed:
resoponding to comments

HarshiniGirish committed:
Merge branch 'develop' into rajat/BIOMASS_data_access

omshinde commented:
@hrodmn : @HarshiniGirish Addressed your comments and the PR is up for re-review. Thanks
